airbyte-cdk/python/airbyte_cdk/sources/declarative/migrations/legacy_to_per_partition_state_migration.py

```diff
@@ -0,0 +1,81 @@
+# Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+
+from typing import Any, Mapping
+
+from airbyte_cdk.sources.declarative.interpolation.interpolated_string import InterpolatedString
+from airbyte_cdk.sources.declarative.migrations.state_migration import StateMigration
+from airbyte_cdk.sources.declarative.models import DatetimeBasedCursor, SubstreamPartitionRouter
+
+
+def _is_already_migrated(stream_state: Mapping[str, Any]) -> bool:
```

Comment:
Can we move this method to the `StateMigration`? Since it's pretty obvious check, and should be the part (in my opinion) of `should_migrate()` logic. 
Maybe there are some future non-obvious benefits having this method outside?

Reply:
this is a good question. I'd prefer we keep StateMigration a pure interface so it's easier to use. The problem with inheriting from a class that has implementation is the developer needs to know when to call super, and what effect it'll have.

I am however open to adding more structure if patterns like these emerge. For now, I don't see the need in forcing the developer to implement both `should_migrate` and `is_already_migrated`